### ISSUE TITLE
API for serverless access?

### ISSUE THREAD

**Issue by JamesPerreault** [sqlitebrowser/dbhub.io#132]
The feature list shows that there is an API for "serverless" access.  I can find no documentation on that.  Is that feature implemented?

**Comment by justinclift** [comment]
@JamesPerreault Not yet.  It's the main thing I want to get done next.

Haven't really designed it out though (eg the first proper steps).

If you're ok for sharing, what sort of stuff are you looking for?  I'll likely use that as the initial use case, to make sure that works well, then expand out from there. :smile:

**Comment by JamesPerreault** [comment]
First, let me give you my use case.  I want to port a desktop recipe application to a web enabled application.  The application's dbackend is a sqlite database.

So at the minimum I would need a query interface.  An sql like interface would probably work. (I.e., I specify the select statement, the results are returned; possibly in chunks).

Updates would be handy, but I think I could use the existing command line tool for that.

**Comment by justinclift** [comment]
Awesome. :smile:

Thinking that through (out loud)...

Being able to execute arbitrary (read only) SQL (eg `SELECT` statements) should be pretty easy.  The server side code for doing the core piece of that is already written and working decently well for our recent visualisation feature.

It should be fairly straight forward to make a "version 1" api using that.  Would you be for trying it out, likely this weekend?

---

Thinking about updates, two main potential approaches spring to mind:

1. Uploading a new database (or database revision)
2. Running (say) `INSERT`, `UPDATE` or `DELETE` statements

It should be ~reasonably easy to add support for uploading new databases, and/or new revisions of an existing one.  That's also code which already exists, so could be leveraged to suit in a timely manner.

Running SQL statements that modify a database (eg `INSERT`, etc) would probably be a lot more tricky, as the entire server side is version controlling everything at present.

So we'd either need to figure out a way to store databases in a "live" (non-version-controlled) fashion... or automatically create new revisions each time an "updating type of statement" is run.  Doable, but it could be fairly complex to make happen.

If you're ok with the first approach (uploading complete new database), via the API, that's probably the one to go for initially.

Thoughts?

**Comment by JamesPerreault** [comment]
Yes, I would be able to try out version 1 of the api this weekend.

As for updates, option 1 works for me.

Thanks!

**Comment by justinclift** [comment]
Excellent.  Will get that happening.

Out of curiosity, do you do much API programming stuff?

Also, what's the primary programming language(s?) and/or frameworks(s?) you're likely to use?

All that info is optional of course, it's just I'll probably do a bit of reading up on it to try and make sure things "work easily". :smile:

**Comment by JamesPerreault** [comment]
I'll be using Node.js / Javascript for the query interface.   The thick app is written in Python, so inserts and updates will likely be in that language. But I'm still thinking about that part.

I have not done any API development, although I do know C/C++.

**Comment by justinclift** [comment]
No worries.  We'll get something working for this weekend, and can iterate on it until we're both happy. :smile:

**Comment by chrisjlocke** [comment]
Back in March (that seems months ago now ... when it was cold and windy...) I was doing a lot of API work in VB (both XML and JSON data) so happy to look at this.  JSON is preferred...
But yes, can send/receive API requests to help with this.

**Comment by justinclift** [comment]
@JamesPerreault Working on getting a basic query interface happening now. :smile:

**Comment by justinclift** [comment]
Well, that turned out to be a lot more fiddly to get the pieces in place than I was expecting. :wink:

That being said, there's a very (**very**) bare bones API server running that people can query their own databases (and any public ones) one.

The server name is `https://api.dbhub.io`.  As an initial starting point, it only does one thing... run a given SQL query on the desired database.  There's no upload, delete, "get properties", "get list of databases", (etc).

For running a SQL query, the address is `https://api.dbhub.io/v1/query` and there are four parameters:

* `dbowner` - The owner of the database to query
* `dbname` - The name of the database to query
* `commit` - (optional) If you want to run the query on a specific database version rather than whichever is the latest, pass in it's commit ID via this parameter
* `sql` - The SQL query to be run, base64 encoded

As a useful way to test it, the command line utility `curl` seems to work well.  Here's an example for running a query against one of the public example databases in my account:

```
curl -F apikey="YOUR_API_KEY_HERE" -F dbowner="justinclift" -F dbname="Join Testing.sqlite" \
  -F sql="U0VMRUNUIHRhYmxlMS5OYW1lLCB0YWJsZTIudmFsdWUKRlJPTSB0YWJsZTEgSk9JTiB0YWJsZTIKVVNJTkcgKGlkKQpPUkRFUiBCWSB0YWJsZTEuaWQ7"  \
  https://api.dbhub.io/v1/query
```

You can get an API key by logging into the DBHub.io website interface as normal, then going into the user Settings page (top right corner after logging in).  There should be an obvious "Generate API key" button available now. :wink:
(note that there's not yet any way to delete or disable keys.  That's a task for another day)

Running the above SQL query **should** return a result as JSON.  The structure it's being returned will **probably** need adjusting though, as it's just something I threw together ages ago for our web interface to use.  I'm happy to change the output format/structure (etc) if needed.

Thoughts?

**Comment by justinclift** [comment]
Oh, forgot to mention... the SQL query in the above example (prior to base64 encoding it) is:

```
SELECT table1.Name, table2.value
FROM table1 JOIN table2
USING (id)
ORDER BY table1.id;
```

**Comment by chrisjlocke** [comment]
Well done!  Nice!  I'll have a go at this - currently doing some API gubbins stuff, so ...

**Comment by chrisjlocke** [comment]
Whatever I throw at the URL, I get this error.

`curl: (35) schannel: next InitializeSecurityContext failed: Unknown error (0x80092012) - The revocation function was unable to check revocation for the certificate.`

**Comment by JamesPerreault** [comment]
I tried it out.  It works find with curl but when I try to do it in a web page I get a CORS error:

> _Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://api.dbhub.io/v1/query. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing)._

Can you make it CORS compliant?   The Content type looks okay, so I think the header is all that is needed.

I think I can work with the existing JSON structure.  If I run into issues, 'll let you know.

Also, my database has images, stored as BLOBS.   I am developing without using them, but at some point I would like to use them.  (Ideally, returned in a base64 encoded string).

 Thanks!

**Comment by justinclift** [comment]
> The revocation function was unable to check revocation for the certificate.

@chrisjlocke Try using curl with the `-k` option.  That tells it to skip HTTPS certificate validation, which sounds like the piece that's failing.

The webserver is using a standard Lets Encrypt certificate, so there *shouldn't* (in theory) be issues with it.  Is the very of curl you're using pretty old?

**Comment by justinclift** [comment]
> Can you make it CORS compliant?

Good point.  I've been able to avoid learning how CORS works so far.  Looks like it's time to get that done.  I'll have a go at it tonight after work. :smile:

With the BLOBs, no worries, that makes sense.  It shouldn't be too hard to get working. :smile:

**Comment by JamesPerreault** [comment]
Actually, I've decided I don't want my credentials on the web page in the clear.  By making the db queries server side, instead of in the browser on the client side, I can avoid that.   I've gotten that working, so I'm good to go.   You can work on the CORS stuff if you want, but I don't think I will be needing it.

Thanks!

**Comment by chrisjlocke** [comment]
> Try using curl with the `-k` option. That tells it to skip HTTPS certificate validation, which sounds like the piece that's failing.

Doh!  Didn't think of updating Curl.  Doing a -V did imply it was an old version.  Downloaded a newer one, and the error was indicating it didn't like the self-signed certificate.  Using the -k option solved that.  Thanks for the hand-holding! 👍 
Works perfectly.

**Comment by justinclift** [comment]
Er... it's **not** using a self signed certificate. :wink:

**Comment by justinclift** [comment]
@JamesPerreault Cool.  I'll skip the CORS stuff for now then.  I'll take a look at the blobs, and see if there's a good way to return them in base64 format.

It won't be tonight though, I'm feeling pretty tired and want to hit the sack early instead. :wink:

**Comment by JamesPerreault** [comment]
I'm having an issue where I cannot query on numeric fields.   Here is an example:

`select unit,amount,rangeamount,item,optional,inggroup,position      from ingredients WHERE deleted == FALSE AND recipe_id == 24;
`
Here's what it looks like bse64 encoded:
`
c2VsZWN0IHVuaXQsYW1vdW50LHJhbmdlYW1vdW50LGl0ZW0sb3B0aW9uYWwsaW5nZ3JvdXAscG9zaXRpb24gICAgICBmcm9tIGluZ3JlZGllbnRzIFdIRVJFIGRlbGV0ZWQgPT0gRkFMU0UgQU5EIHJlY2lwZV9pZCA9PSAyNDs=`

I've no issues querying on string fields.

Ideas ?

**Comment by justinclift** [comment]
Thanks, investigating now. :smile:

**Comment by justinclift** [comment]
Ahhh, hang on.  I think it's just a standard SQL error.  This is showing up in the logs:

```
Error when preparing statement for database: near "'24'": syntax error (select unit,amount,rangeamount,item,optional,inggroup,position      from ingredients WHERE deleted == FALSE AND recipe_id  '24';) (SQL logic error)
```

Probably needs a single `=` symbol, rather than doubling them up?

Also, I'll see if there's a way to pass back error messages like this in a useful fashion. :smile:

**Comment by justinclift** [comment]
Hmmm, the `==` sign approach works in the `sqlite3` command line shell.  Interesting.  It should work for the SQL query then.

I'll probably need to investigate this after work though, as I need to get ready for that now. :wink:

**Comment by justinclift** [comment]
k, I've replicated the problem in my local development environment, and will get back to figuring out a solution after work tonight. :smile:

**Comment by justinclift** [comment]
This is a weird bug.  While this query works:

```
select unit, amount from ingredients WHERE deleted = FALSE AND recipe_id = 24
```

This one doesn't:

```
select unit, amount, rangeamount from ingredients WHERE deleted = FALSE AND recipe_id = 24
```

But only when called via the SQLite driver we're using.  Both queries work fine from the command line.  Am debugging now...

**Comment by justinclift** [comment]
@JamesPerreault This should be fixed now.  It turned out to be a dumb bug on my part, where I'd blindly left the code ignoring any result rows which contain a null value.

On the plus side, there shouldn't be **too** many more dumb bugs around.  Hopefully. :smile:

**Comment by JamesPerreault** [comment]
It's working!   There is one bug, where null values are being returned as `"<i>NULL</i>"` .   Not a big deal, as I can check for it and set it appropriately ( right now "" ).   With that work around, my code is working.

Thanks for the quick work!

**Comment by justinclift** [comment]
Awesome! :smile:

For the NULL values, what would be better to return?  Empty string, or something else, or ?

I'll take a look at getting the blobs-as-base64 piece done, probably tomorrow too. :smile:

**Comment by justinclift** [comment]
As a data point, that `<i>NULL</i>` is coming from this statement:

https://github.com/sqlitebrowser/dbhub.io/blob/14e1a31bf9bcf9e45ba922c61c0f75b4b567d2b5/common/sqlite.go#L875-L876

It should be pretty simple to change. :smile:

**Comment by justinclift** [comment]
BLOB values are handled just above that code fragment too, currently like this:

https://github.com/sqlitebrowser/dbhub.io/blob/14e1a31bf9bcf9e45ba922c61c0f75b4b567d2b5/common/sqlite.go#L864-L865

**Comment by MKleusberg** [comment]
> For the NULL values, what would be better to return? Empty string, or something else, or ?

Since we're using JSON, I'd just use ```null``` for NULL values. This allows the user to differentiate NULL values and empty strings.

**Comment by justinclift** [comment]
@MKleusberg As in "null" as a direct text string?

Can make that change right away, unless you want to do it?

**Comment by MKleusberg** [comment]
```null``` as in
```json
{"Name":"Name","Type":3,"Value":null}
```

I'm currently looking through some other code. So just go ahead if you want to :smile:

**Comment by justinclift** [comment]
@JamesPerreault The server had the "ignore rows with binary data in them" option turned on, just like it was turned on for nulls before.  That's now been turned off, so you might want to double check and make sure things are still ok on your end. :smile:

@MKleusberg I'll look at the null representation tomorrow.  I need to figure out what internal representation to use in the data so it gets turned into `null` when the Go json.Marshall() function is called.  It's probably super simple, but I'm yawning my head off so it's probably better I get sleep now instead. :wink:

**Comment by MKleusberg** [comment]
@justinclift The correct internal representation apparently is ```nil```. I'm not sure about any side effects on the webui though. So I'll leave that to you :wink:

**Comment by JamesPerreault** [comment]
It's still working, as far as I can tell.

**Comment by justinclift** [comment]
@MKleusberg Just realised that yep, it's [super simple](https://github.com/sqlitebrowser/dbhub.io/commit/6f11766ccfae2fc6a32cd3408af84f57efdc0eaf).  So made the change just now and restarted the server. :smile:

@JamesPerreault The output for nulls should be correct now, as per JSON spec.  In theory. :angel:

**Comment by justinclift** [comment]
@MKleusberg Side effects in the webUI is a good point, as the visualisation code is using the same function.  So we'd better check that.  Not tonight though. :wink:

**Comment by justinclift** [comment]
@JamesPerreault Let me know when you'll be up for doing stuff with base64 encoded images.  My suspicion is it'll be pretty simple to do the base64 encoding server side, but I have no idea what you'll need to do client side for it to be shown as an image.

**Comment by JamesPerreault** [comment]
I'm a bit busy at the moment, so it probably won't be until middle of next week.

**Comment by justinclift** [comment]
@JamesPerreault No worries. :smile:

**Comment by justinclift** [comment]
Just added initial support for base64 encoding the returned BLOB (binary) fields.  Hopefully that turns out to be useful. :smile:

For good measure, also added initial docs of the available API end points (eg `/v1/query`), parameters, etc to the API server.  They're viewable by just going to the root directory in a browser.  eg: https://api.dbhub.io

**Comment by chrisjlocke** [comment]
Talking of 'docs' and stuff, should we/can we/may we use Swagger?
https://swagger.io/tools/

Don't mind dabblin' with this, as its interesting to me.

**Comment by justinclift** [comment]
@chrisjlocke Go for it.  Seen lots of mention of Swagger, but never really took the time to grok it. :smile:

**Comment by JamesPerreault** [comment]
You can display base64 encoded images in html, by setting the image _src_ tag to:
`src="data:image/jpeg;base64, <base64 encoded image"`

or whatever the image type is.

I just tried it out, and it works beautifully!

Thanks again for the quick work.

**Comment by justinclift** [comment]
@JamesPerreault Awesome! :smile:

Is your recipe application something we can point people at as a working example?

**Comment by JamesPerreault** [comment]
Yes.  I haven't put it online yet, as all my development  has been local.  But as soon as I do, I'll let you know.  It might be a few days.

**Comment by justinclift** [comment]
Cool.  And no worries at all. :smile:

**Comment by justinclift** [comment]
Note to self - add some unit tests for the api. :smile:

**Comment by chrisjlocke** [comment]
I'll experiment more with this next week ... currently working on JSON data at the moment anyway - hopefully Monday.  I'll get a WikiHow tutorial going! 😄

**Comment by justinclift** [comment]
That'd be super welcome! :grin:

**Comment by chrisjlocke** [comment]
There are different security levels - OAuth, etc.  This is slightly above my head already, but I assume we've already got OAuth stuff in place?

![image](https://user-images.githubusercontent.com/7872218/87871746-047b4700-c9ab-11ea-8fb9-27ffedf4fdfb.png)

I guess I'll play with 'simple API' first and worry about security later... 😉   I assume it can be easily changed and tweaked.

**Comment by justinclift** [comment]
Nope, no OAuth in place for the API access.  I don't know much about OAuth, kind of assuming (perhaps wrongly) it's an interactive-login thing.

Yeah, the "Simple API" option seems like a reasonable things to investigate first. :smile:

**Comment by justinclift** [comment]
As a general data point, it'd probably be useful for us to implement enough functionality with the API, that someone could add a "DBHub.io" driver to SQLAlchemy:

&nbsp; &nbsp; https://www.sqlalchemy.org

That's a **very** widely used (Python based) data access library, which already supports "apikey" based access to remote (server) databases.  We could probably copy + adapt one of the existing drivers there, plus add the needed API end points to our system.  Then contribute the driver to the SQLAlchemy project, so it's fairly widely available.

**Comment by justinclift** [comment]
As another "note to self", it should be pretty trivial for me to add a working example of calling the new API from Go. :wink:

**Comment by JamesPerreault** [comment]
Here's a link to my repository  https://github.com/JamesPerreault/MyGourmetRecipes  .  Because of possible copyright issues with the recipes themselves, I'm not including a link to the active page.   Instead, I included a few screenshots.

**Comment by justinclift** [comment]
Awesome!  Thanks heaps @JamesPerreault.

I'll add a link to that from the API docs (likely this weekend), so people have a "fully worked" reference example. :smile:

**Comment by justinclift** [comment]
@JamesPerreault Just added a link to your example application from the main api.dbhub.io page.  Hopefully people find it useful. :smile:

**Comment by justinclift** [comment]
As a data point, I've started creating a Go library which people can use to access their databases directly:

&nbsp; &nbsp; https://github.com/sqlitebrowser/go-dbhub

It's extremely early stage, but does run remote queries and return the results as would be expected.

The API server still needs a bunch more functionality added, and this Go library should be useful for testing it.

* Function for returning a list of available databases
* Function(s) for listing tables and views in the remote database
  * Initial code written that returns *both* tables and views.
  * No docs nor examples written
  * No tests written
* Function for downloading complete database
* Function for uploading complete database
* Function for retrieving database details (size, branch, commit list, whatever else is useful)
* And create a reasonable example application written in Go




Need to add more functionality both to the API server and this

**Comment by justinclift** [comment]
Small update... tab characters are now allowed in the SQL queries being passed through.  eg:

```
SELECT foo
FROM bar
WHERE abc = 1
	AND def = 2
	AND ghi = 3;
```

Found and fixed that problem a few days ago, when trying out initial non-trivial SQL queries. :wink:

**Comment by justinclift** [comment]
Does anyone have a minute to look through the initial API docs?

&nbsp; &nbsp; https://api.dbhub.io

@MKleusberg have been adding what (hopefully :wink:) is reasonable initial API functionality over the last few days.  But we haven't updated any docs until today.

The general look and feel seems (to me) like it should  workable, but I'm not 100% sure.  And I haven't had to use them from the perspective of "someone who wants to use the API".

Oh, most of the "Return values" still have "TBD" (To Be Determined) as a placeholder for now.  That still needs doing and I'll get to it either tonight or tomorrow. :wink:

**Comment by JamesPerreault** [comment]
Is there a reason all the return values say TBD? That's a bit confusing.

Thanks!

Jim

**Comment by chrisjlocke** [comment]
Thats probably due to this:

> Oh, most of the "Return values" still have "TBD" (To Be Determined) as a placeholder for now. That still needs doing and I'll get to it either tonight or tomorrow. 

We've tried giving Justin two keyboards to type faster, but that didn't work out so well.  So now we just throw tennis balls at him if he sleeps for too long.
He'll be back fixing this soon.  We've got to give him /some/ sleep.

**Comment by justinclift** [comment]
Yeah, it's just a sign of the "work in progress" nature of the API docs.  I tend to commit and publish stuff "live" while I'm working on it, except where it's likely to cause a security (or similar) problems. :wink:

The `TBD` placeholders will probably be replaced with the correct info today and tomorrow.

Oh, I'd like to simplifiy the very verbose format of results from `/v1/query` too.  Probably to something that's just a 2 dimensional array of the output values, with maybe the column names in the first row.  Something akin to a JSON-ified CSV file, though I haven't really thought it through yet.  er... TBD. :wink:

**Comment by justinclift** [comment]
**Most** of those TBD's have been filled out now.  There are a few pieces that still need to be done (two fields for `Column`, all of the `Diffs` fields, and `Webpage` function needs adding).

I'll probably get those done tomorrow sometime. :smile:

**Comment by justinclift** [comment]
@JamesPerreault Adding upload functionality to the API today, if that's of interest. :smile:

It took a bit longer than expected, as there turned out to be a bug with the nightly builds of DB4S upload functionality (introduced a few weeks ago).  That bug meant a first or initial upload worked, but subsequent ones didn't (without a manual workaround).  That bug is fixed now though, so uploads work again in the nightly builds. :angel: 

With the upload functionality added, it should be possible for people to build their own custom UI's for doing things.  The API now has the same functionality which DB4S itself uses.

The "Diff" functionality still needs documenting, which I'll probably do later on today too.  I need to go through that code in depth and better understand how it works. :smile:

**Comment by MKleusberg** [comment]
@justinclift I don't have enough time to write documentation for the diff call right now but this might help you if you want to write it. It returns a JSON object like this:
```js
{
  "diffs": [
    // Change sets are listed here
  ]
  // More attributes might come later
}
```

Each change set looks like this. There is one change set object per changed object (table, ...) in the database. So a thousand changed rows in a table still produce only one such object. A single changed row in 5 different tables produces five such objects.
```js
{
  "object_name": "Name of the changed table/index/view/trigger",
  "object_type": "table", // or "index", or "trigger", or "view",
  "schema": ...,  // null when the schema hasn't changed
  "data": ...  // null when the data hasn't changed. Always null for object types other than "table"
}
```

If the ```schema``` attribute is not null, it is an object like this:
```js
{
  "action_type": "add",  // or "modify", or "delete". "add" means the object was created, "modify" means its definition changed, and "delete" means it was dropped.
  "sql": "..." // The SQL statements needed to make the schema change happen. Empty when merge strategy is set to "none"
}
```

If the ```data``` attribute is not null, it is an array of objects like this. There is one object in the array per changed row.
```js
{
  "action_type": "add",  // or "modify", or "delete". "add" describes inserted rows, "modify" updated rows, and "delete" deleted rows
  "pk": [{"name": "Name of the PK column", "type": "Type of the PK value", "value": "PK value"}], // For tables with multiple PK columns the array contains more than one item. These values tell you the key of the row which changed. If you need the old and/or new data you can use this key to query the old and/or new database
  "sql": "..." // The SQL statements needed to make the schema change happen. Empty when merge strategy is set to "none". The exact value depends on the merge strategy.
}
```

**Comment by justinclift** [comment]
Thanks @MKleusberg, that'll be a good starting point.  My intention is to start hooking up the `Diff()` code to our Merge Request pieces after I get this initial set of API functions working.  Hopefully this week.

So I need to step through the diff code to properly understand it anyway. :shrug:

**Comment by MKleusberg** [comment]
Awesome :smile: Looking forward to that! Maybe two more notes on that:
* If you want to show the differences between two revisions or branches, set the merge strategy no ```NoMerge``` to save CPU time
* If you want to implement actual merging, you'll need to set the merge strategy as described in [this commit message](https://github.com/sqlitebrowser/dbhub.io/commit/17d37a1e4b78eebb1227f49b134a5a0db1da5467).

**Comment by justinclift** [comment]
As a future reference point, the [OpenAPI](https://www.openapis.org) project is probably something we'd want to investigate for a future DBHub.io API version.

**Comment by justinclift** [comment]
As a data point, I've just "enabled" CORS for all API requests.

Saying "enabled" in quotes there, as all I did was blindly add an `Access-Control-Allow-Origin: *` header to all API responses.

Mainly because that was super easy, rather than investigating CORS in any depth whatsoever.

**Comment by JamesPerreault** [comment]
The other thing needed for CORS is the Context-Type has to be set to
text/plain or something like that. I'll test it out and see if I have any
issues.

Jim

**Comment by justinclift** [comment]
Ahhh cool.  The Context-Type should already be getting set to something appropriate. :smile:

**Comment by justinclift** [comment]
As a general thought, it might turn out to be useful to implement some kind of permissions system for API keys.

Something that would let people do (for example):

* API key "foo" is only allowed to access database "bar".  And it's limited to just the "/v1/query" call.
* API key "baz" is only allowed to access database "abc".  It's allowed all read-only calls.  (eg no upload, no delete)
* API key "stuff1" has full access.

Something like that would be one way to limit potential damage of leaked (etc) API keys.

**Comment by JamesPerreault** [comment]
That would be handy. I could put a read only key in a web page, and not
have to proxy things through AWS like I am now. But even without that, it
would be great protection against SQL injection attacks.

Jim

**Comment by justinclift** [comment]
From an implementation perspective, it should be reasonably straight forward to add the UI pieces to the Settings page, so people can associate a given API key with a given database, and/or make them read only.

With the SQL injection attack consideration (specifically), things *should* already be in a decent place. I was careful to implement the "[Defense Against Dark Arts](https://www.sqlite.org/security.html)" recommendations, and also limit the incoming queries so only `SELECT` statements are run.

**Comment by justinclift** [comment]
@JamesPerreault I'll probably get back into finishing the API piece off, then implementing initial permissions next week, or on the weekend after.  I'm just a bit burnt-out at the moment, and need to not really look at this stuff for a few days. :wink:

**Comment by justinclift** [comment]
"Next week" turned out to be a bit ambitious.  Needed to take an extended break after burn-out. :wink:

Anyway, started experimenting with this stuff again after work.  Initial concept mock up (no actual code behind it):

![2021-06-01-api_keys_concept_layout_v1](https://user-images.githubusercontent.com/406299/120227329-4c861e00-c28c-11eb-8993-64ea0d81ea26.png)

* That table with the function call on/off toggles will probably need to be moved to a separate page, otherwise people with multiple API keys will have a *long* page to display. :wink:
* Each API key could have a specific database chosen (eg from drop down list of all databases), or "all" so the enabled functions can be locked down pretty well
* Once we figure out how to do read/write databases (not just read only, and I have some ideas), we'll probably need to update the list of function calls again.  Or maybe add some extra pieces, around read/write capabilities
* It'd probably be useful to have data shown for when an API key was last used.  Maybe even an audit log or something?

**Comment by justinclift** [comment]
@JamesPerreault As a data point, we started putting a lot of effort into DBHub.io again after the new year.

*Some* of the new stuff might be useful for you. :smile:

Am mostly thinking of the "Live databases" functionality:

&nbsp; &nbsp; https://sqlitebrowser.org/blog/live-databases-are-now-live/

... as it's specifically for people wanting write enabled databases.  eg INSERT/UPDATE/DELETE

That being said, we still need to get the outstanding API permissions stuff done.

That's likely to be right after I get Stripe subscriptions hooked up in our code, which itself is tentatively pencilled in for next week.

**Comment by justinclift** [comment]
No worries.  This new "live databases" stuff is in very (very!) active development, so if you notice rough edges / have suggestions / etc please point them out. :smile:
